src/replication.c

```diff
@@ -1803,14 +1803,21 @@ void updateSlavesWaitingBgsave(int bgsaveerr, int type) {
         if (slave->replstate == SLAVE_STATE_SEND_BULK_AND_STREAM) {
             /* This is the main channel of the slave that received the RDB.
              * Put it online if RDB delivery is successful. */
-            if (bgsaveerr == C_OK)
+            if (bgsaveerr == C_OK) {
+                /* Notify the task that the snapshot bulk delivery is done */
+                if (slave->flags & CLIENT_ASM_MIGRATING)
+                    asmSlotSnapshotSucceed(slave->task);
                 replicaPutOnline(slave);
-            else
+            } else {
                 freeClientAsync(slave);
+            }
         } else if (slave->replstate == SLAVE_STATE_WAIT_BGSAVE_END) {
             struct redis_stat buf;
 
             if (bgsaveerr != C_OK) {
+                /* Notify the task that the snapshot bulk delivery failed */
+                if (slave->flags & CLIENT_ASM_MIGRATING)
+                    asmSlotSnapshotFailed(slave->task);
```

Comment:
is it possible we collect rdb status after we free the task?

Reply:
we set `c->task = NULL` if task is failed.
controlling with the CLIENT_ASM_MIGRATING seems not a right idea, i revert it. since we may log the client as a replica.

but i just realized i ignored the rdb channel client when the task is done normally.
generally the destination will close the rdb channel connection and then stream buffer into db, on the source, it may take some while to make this take done, so it won't cause use-after-free.

but we can reset `task` of rdb channel when task is going normally

Reply:
resolved in https://github.com/tezc/redis/pull/62/commits/9b51c33bc9a7578754970ff3d2f600d5fb2bfbf6